oo refactor

andrewkrug opened the pull request:
This refactor of the CIS library implements basic nullObject pattern to improve test-ability of code and lays the foundation for signatures.  

Signatures are not currently enforced all the way through but all objects support them. 

Here's a code sample of publishing to the new stream using the new publisher object. 

This PR also introduces a simplified processor object that collapses the logic in cis_functions. 

```
    event_file = os.path.join(
        os.path.abspath(os.path.dirname(__file__)), 'event.json'
    )

    with open(event_file, 'r') as event_data:
        profile_data = (event_data.read().encode('utf-8'))


    from cis.libs import encryption

    encryptor = encryption.Operation(boto_session=boto3.Session(region_name='us-west-2'))

    event = {
        'publisher': {'id': 'mozillians.org'},
        'signature': {},
        'profile': encryptor.encrypt(profile_data)
    }

    print(event)

    res = handle(event, None)
```

andrewkrug commented:
@gdestuynder @johngian didn't have time to pep8 all the things.  I can make another pep8 pass post sprint review.

johngian commented on cis/__init__.py:
```diff
@@ -1,7 +1,3 @@
 __all__ = [
```
Either remove or fix this entry. This way when importing everything under module ``cis`` imports nothing.

andrewkrug replied:
fixed

gdestuynder commented on cis/libs/streams.py:
```diff
@@ -0,0 +1,57 @@
+        json_payload = json.dumps(base64_payload).encode('utf-8')
+        return json_payload
+
+
```
lotsa straight LF ;-)

andrewkrug replied:
Done

johngian commented on cis/libs/connection.py:
```diff
@@ -0,0 +1,52 @@
+            raise(e)
+
+    def connect(self):
+        if self.connection_type is None:
```
``connection_type`` can only be ``None`` if user selects ``None`` as an argument. By default it is a required argument in the ``__init__`` signature and if it's not supplied python will raise an error anyway.

andrewkrug replied:
This was also intentional to support a test case.  However, this has now been removed as well.

gdestuynder commented on cis/libs/validation.py:
```diff
@@ -0,0 +1,67 @@
+                    else:
+                        pass
+                except Exception as e:
+                    print(e)
```
s/print/traceback.print_exc/ ? (or stray debug code?)

andrewkrug replied:
I elected to drop in loggers here. I'm all for improving the bubble up to CloudWatch so let's decide on the best way and make all the exception logging support that.

gdestuynder replied:
@andrewkrug i think the print is still in there )probably other places too)
using traceback gives you the full traceback instead of just the exception name which is generally very helpful when debugging.

I would recommend dropping all print() outside of tests and either adding or enhance logger code to log the exception traceback wherever there is a catchall (ie `except Exception as e: ...`)

johngian commented on cis/libs/connection.py:
```diff
@@ -0,0 +1,52 @@
+            raise AttributeError(
+                "Could not determine connect type.  Set client or resource."
+            )
+        elif self.connection_type == "client":
```
Instead of having all these ``if/elif/else`` cases and check against strings I suggest we have a default connection_type that works in most cases and override it with a ``boto`` client object instance/resource/session.

andrewkrug replied:
I'm only currently passing around boto3 session objects.  So I've left the code as is but made it default to session as the default type.

gdestuynder commented on cis/plugins/validation/mozilliansorg_publisher_plugin.py:
```diff
@@ -36,8 +36,12 @@ def run(publisher, user, profile_json):
 
-    # Check for any non-whitelisted attribute modification
-    # Note that no extra attribute may be added at this stage as we already performed schema validation
+    # Check the easiest case.  None type.
```
this means we need to preimport users, until we have publishers which are allowed to create new users (workday/auth0 i believe)

andrewkrug replied:
We need to talk about this a bit.

gdestuynder replied:
@andrewkrug  I will be unavailable for most of next week - my vote would be to let mozilliansorg create users when they dont exist only, and until auth0/workday publishers exist. Happy to talk tomorrow though.

johngian commented on cis/libs/connection.py:
```diff
@@ -0,0 +1,52 @@
+        try:
+            # Attempt to init default boto session and raise if fails
+            boto3.setup_default_session(profile_name=self.profile)
+        except Exception as e:
```
Catch a specific exception here. This way you are catching  and re-raising all possible exceptions which is equivalent of not having this try/except at all.

andrewkrug replied:
This was intentional to support a desktop use case.  But... I've eliminated the block now so it will just blow up if no boto profile exists.

gdestuynder commented on cis/processor.py:
```diff
@@ -0,0 +1,127 @@
+        elif self.stage == 'streamtoidv':
+            res = self._vault_stage()
+        elif self.stage == 'idvtoauth0':
+            # TBD fold in the authzero logic from idvtoauth0 in cis_functions
```
nit: s/TBD/FIXME. i usually just go with smth like 'XXX FIXME TBD INSERTURSHERE' so that everyone's favorite IDE picks these up

gdestuynder replied:
would also add that to the libs/authzero.py placeholder

andrewkrug replied:
didn't do this yet.

johngian commented on cis/libs/encryption.py:
```diff
@@ -0,0 +1,98 @@
+
+    def _get_data_key_from_kms(self, encryption_context={}):
+        """
+        
```
Please remove empty line here.

andrewkrug replied:
Done

johngian commented on cis/libs/encryption.py:
```diff
@@ -0,0 +1,98 @@
+
+        try:
+            self.kms = boto_session.client(service_name='kms')
+        except Exception as e:
```
Same here. Be more specific on the exceptions you catch.

gdestuynder commented on cis/processor.py:
```diff
@@ -0,0 +1,127 @@
+            # TBD fold in the authzero logic from idvtoauth0 in cis_functions
+            pass
+        else:
+            # Unhandled pass for anything not handled in block.  Basically yield to block.
```
logger.exception('unhandled processor stage type {}'.format(self.stage)) or similar?

andrewkrug replied:
Agree... but let's get this in a logger pass.

johngian commented on cis/libs/encryption.py:
```diff
@@ -0,0 +1,98 @@
+        try:
+            self.kms = boto_session.client(service_name='kms')
+        except Exception as e:
+            self.kms = None
```
What's the point of having ``kms`` None if you are going to use ``None`` later. ``self.kms`` should never be none. And if ``boto_session.client()`` fails it should raise an error.

andrewkrug replied:
This is a fall through to avoid having to patch / mock KMS.  I've left the code as is but improved the exception handler.

gdestuynder commented on cis/processor.py:
```diff
@@ -0,0 +1,127 @@
+        self.decrytped_profile = json.loads(self._decrypt_profile_packet())
+
+        if self.stage == 'validator':
+            logger.info('Validator logic activated.')
```
nit: seems like logger.debug to me, or actually, move that one 1 line up and:
logger.info('Attempting to load stage processor logic: {}'.format(self.stage)) instead

andrewkrug replied:
Same as above.... let's agree on structured log format and then make another paired pass through the code base until we're all happy with the logging.

johngian commented on cis/libs/streams.py:
```diff
@@ -0,0 +1,57 @@
+import base64
+import json
+from cis.settings import get_config
```
Please add a line to seperate python/3rd party/local imports.

andrewkrug replied:
Done

gdestuynder commented on cis/publisher.py:
```diff
@@ -0,0 +1,136 @@
+        try:
+            ChangeDelegate.__init__(self, publisher, signature, profile_data)
+        except Exception as e:
+            print(e)
```
s/trackeback.print_exc/

johngian replied:
Let's avoid printing, that's the reason we use ``logger`` instead.

johngian commented on cis/libs/streams.py:
```diff
@@ -0,0 +1,57 @@
+        :partition_key: Kinesis partition key used to publish data to
+        """
+
+        if self.kinesis_client is None:
```
The way that you have coded it ``self.kinesis_client`` is always ``None`` when running this method. Please move the initialization to ``__init__``.

andrewkrug replied:
It's initialized opportunistically if not passed in with the object constructor.  This makes it easy to use moto in the test pattern.

johngian commented on cis/libs/streams.py:
```diff
@@ -0,0 +1,57 @@
+        }
+
+        stream_arn = self.config('kinesis_stream_arn', namespace='cis')
+        stream_name = stream_arn.split('/')[1]
```
Since we are refactoring this, let's avoid ``split`` on something that we define ourselves. Let's prepare the correct format in the config level.

andrewkrug replied:
This is an enhancement I'd rather grab after merge.

johngian replied:
Pleaes file a bug then.

gdestuynder commented on setup.py:
```diff
@@ -1,11 +1,25 @@
 
 setup(
     name='mozilla-iam-cis',
     version='0.1dev',
```
probably s/0.1dev/1.0/

andrewkrug replied:
unsure how we track versions of this.

johngian commented on cis/libs/streams.py:
```diff
@@ -0,0 +1,57 @@
+
+        json_payload = json.dumps(base64_payload).encode('utf-8')
+        return json_payload
+
```
Please remove trailing empty lines.

andrewkrug replied:
Done

johngian commented on cis/libs/validation.py:
```diff
@@ -0,0 +1,67 @@
+import os
+import logging
+from pluginbase import PluginBase
```
Please add a line to separate python dependencies from 3rd party.

gdestuynder commented on tests/old/1_test_encryption.py:
```diff
@@ -0,0 +1,81 @@
+            EncryptionContext={}
+        )
+
+"""
```
not currently working? (test_decrypt_payload)

andrewkrug replied:
Fixed.

johngian commented on cis/libs/validation.py:
```diff
@@ -0,0 +1,67 @@
+        """Source of truth for all validation options."""
+        s = Schema(self.publisher, self.profile_data, self.user)
+
+        if s.validate() is True:
```
This is equivalent of ``return s.validate()``.

johngian commented on cis/libs/validation.py:
```diff
@@ -0,0 +1,67 @@
+        else:
+            return False
+
+class Schema(object):
```
Please move this before ``class Operation`` to improve readability.

andrewkrug replied:
Done

johngian commented on cis/libs/validation.py:
```diff
@@ -0,0 +1,67 @@
+                        os.path.abspath(
+                            os.path.dirname(__file__)
+                        ),
+                        '../plugins/validation/'
```
Let's avoid relative paths. I suggest we add a config entry the returns an absolute path. Because:

gdestuynder replied:
@johngian seems your comment was cut

andrewkrug replied:
Let's file a bug for enhancement on this issue.

johngian commented on cis/plugins/validation/json_schema_plugin.py:
```diff
@@ -7,12 +7,13 @@
     # Check for content validity against jsonschema
     jsonschema_validate(profile_json, cis_schema)
-    return True
+
```
Please remove these new empty lines that you added.

andrewkrug replied:
Done

johngian commented on cis/processor.py:
```diff
@@ -0,0 +1,127 @@
+import os
+
+from cis import user
+
```
Remove this line.

andrewkrug replied:
Done

johngian commented on cis/processor.py:
```diff
@@ -0,0 +1,127 @@
+            ciphertext_key=self.encrypted_profile_data.get('ciphertext_key'),
+            iv=self.encrypted_profile_data.get('iv'),
+            tag=self.encrypted_profile_data.get('tag')
+
```
Remove empty line.

andrewkrug replied:
Done

johngian commented on cis/processor.py:
```diff
@@ -0,0 +1,127 @@
+    def run(self):
+        # Determine what stage of processing we are in and call the corresponding functions.
+
+        self.decrytped_profile = json.loads(self._decrypt_profile_packet())
```
Please fix typo and its references.

andrewkrug replied:
Unsure what the typo is here... Left as is.

andrewkrug replied:
I see it now... D'oh.

johngian commented on cis/processor.py:
```diff
@@ -0,0 +1,127 @@
+
+        self.decrytped_profile = json.loads(self._decrypt_profile_packet())
+
+        if self.stage == 'validator':
```
Again I would avoid checking against strings and have all these nested ``if/elif/else`` cases which are very error prone and hard to follow. Since you are going the OO way it would be far more elegant to subclass the different stages.

andrewkrug replied:
This is an enhancement.  I know this block of code is a little unwieldy and smells of non-oo design pattern. I could collapse this into sub-classed objects now.  But it works now... So let's merge and then enhance it together until it's what we want.

johngian commented on cis/processor.py:
```diff
@@ -0,0 +1,127 @@
+
+    def _get_stage(self):
+        # Let the object know what phase of operation we are running in.
+        stage = os.environ.get('APEX_FUNCTION_NAME', None)
```
Let's avoid having cross dependencies of ``cis_functions`` and ``cis``. CIS core is the logic that runs behind the functions. We want to keep them clean and atomic.

By adding this logic here we import code from ``cis`` to ``cis_functions`` that depends on ``cis_functions`` logic to run.

andrewkrug replied:
I'm going to leave this here for now.  But I agree with you at a functional level.  The two should be independent of one another. Let's reach an agreement on how CIS should identify whether it's running in a stream processor in lambda or not.  

FYI I don't see a future where we have stream processors that don't run in lambda.

johngian commented on cis/publisher.py:
```diff
@@ -0,0 +1,136 @@
+import base64
+import json
+
+from cis.libs import connection
```
Please sort imports alphabetically.

andrewkrug replied:
Done

johngian commented on cis/publisher.py:
```diff
@@ -0,0 +1,136 @@
+from cis.libs import connection
+from cis.libs import encryption
+from cis.libs import validation
+
```
Remove empty line.

andrewkrug replied:
Done

johngian commented on cis/publisher.py:
```diff
@@ -0,0 +1,136 @@
+            ChangeDelegate.__init__(self, publisher, signature, profile_data)
+        except Exception as e:
+            print(e)
+            print("returning null object")
```
Same here.

johngian commented on cis/publisher.py:
```diff
@@ -0,0 +1,136 @@
+
+        return response
+
+class Change(ChangeDelegate):
```
This whole class can be truncated to ``ChangeDelegate``. You are just overriding ``__init__`` which if it fails you log something. This adds much more complexity without any added value.

andrewkrug replied:
This is a common pattern for handling testing where you need a guaranteed object out the other side.  I can see us adding more logic to this.  So I'm inclined to leave it in place.

johngian commented on cis/user.py:
```diff
@@ -0,0 +1,66 @@
+"""First class object to represent a user and data about that user."""
+import logging
+from cis.settings import get_config
```
Seperate imports.

andrewkrug replied:
Done

johngian commented on cis/user.py:
```diff
@@ -0,0 +1,66 @@
+class Profile(object):
+    def __init__(self, boto_session=None, profile_data=None):
+        """
+        
```
Please remove empty line.

andrewkrug replied:
Done

johngian commented on cis/user.py:
```diff
@@ -0,0 +1,66 @@
+            return False
+
+    def retrieve_from_vault(self):
+
```
Same here.

andrewkrug replied:
Done

johngian commented on cis/user.py:
```diff
@@ -0,0 +1,66 @@
+        try:
+            response = self.dynamodb_table.get_item(Key=user_key)
+            self.profile_data = response
+        except Exception:
```
Again please be specific in the ``Exception``.

johngian replied:
This way you are hiding the actual error so we don't know and cannot debug what went wrong.

johngian commented on cis/user.py:
```diff
@@ -0,0 +1,66 @@
+        return response
+
+    def store_in_vault(self):
+        # Put data to DynamoDB
```
Use docstrings instead:
```Put data to DynamoDB``.

andrewkrug replied:
Ironically I ported this from what was in master branch.  But I've removed these comments completely now.  I think the method names are explicit enough such that they don't need a comment.

johngian commented on cis/user.py:
```diff
@@ -0,0 +1,66 @@
+            response = self.dyanmodb_table.put_item(
+                Item=self.profile_data
+            )
+        except Exception:
```
Same here.

johngian commented on requirements-dev.txt:
```diff
@@ -0,0 +1,25 @@
+appdirs==1.4.3
```
Why are you renaming this file to ``requirements-dev.txt``?

andrewkrug replied:
In order to include testing-only depends like moto.

johngian commented on setup.py:
```diff
@@ -1,11 +1,25 @@
+from setuptools import setup
+from setuptools.command.install import install as _install
 
+import unittest
```
Please move this to the start of the file. Import order is:

* Complete module imports
* submodule imports

johngian commented on setup.py:
```diff
@@ -1,11 +1,25 @@
+    test_suite = test_loader.discover('tests', pattern='test_*.py')
+    return test_suite
+
+class install(_install):
```
Let's avoid monkeypatching setup utils. This is an antipattern.

andrewkrug replied:
Removed.

johngian commented on tests/old/1_test_encryption.py:
```diff
@@ -0,0 +1,81 @@
+import base64
```
There should be no old and new tests. Please fix naming. Also test names should start with ``test__`` to enable discovery.

andrewkrug replied:
Removed old tests.

johngian commented on tests/old/1_test_streams.py:
```diff
@@ -0,0 +1,95 @@
+import json
```
Same here.

johngian commented on tests/old/1_test_validation.py:
```diff
@@ -0,0 +1,176 @@
+import base64
```
Same here.

johngian commented:
Please fix all PEP8 issues.

johngian commented:
Please squash all commits to one or more atomic commits.

andrewkrug commented:
@johngian @gdestuynder I fixed most of the asks here.  Some of the things I left in place, like the superclassing of the parent objects because they make the cis_functions more testable.  I left the stage environment variable in the code base as well due to the fact that you can either : a. set the environment var or just not import the processor code.  All processors will be apex functions.  This does not effect the ability of non-apex things to publish events.  

Overall I've also cleaned up exception handling and print statements and replaced with loggers.  I'm sure I've missed something though.

andrewkrug commented:
Responded to almost every review request.  I'd like to suggest that we do the following:

A. Accept this as is and then merge the corresponding CIS functions.
B. Reject this pull completely and then pull into a development branch where we all iterate.
C. Reject the attempt completely and go back to the code base the way it was, then spend another full cycle on a refactor.  

I am pretty happy with how things are looking so I would not prefer C.

johngian commented:
There are a lot of replies to comments that suggest fixing stuff in the future. If we want to unblock this PR without them fixed please open up issues.

For example issues that need attention at some point (either now or in the future):

* Logging
* Wrong handling of exceptions
* Subclassing instead of string checking for states
* Env vars for APEX

If there are more feel free to add.
